File: tests/node_table_slab_test.cpp
#include <nanobdd/Bdd.h>
#include <nanobdd/nanobdd.h>

#include <cassert>
#include <iostream>
#include <vector>

using namespace nanobdd;

int main() {
  constexpr size_t tableSize = 5; // intentionally tiny to force growth
  constexpr size_t cacheSize = 5;
  constexpr size_t varNum = 3;

  init(tableSize, cacheSize, varNum);

  const auto baseline = numNodes();

  // Trigger many unique nodes to push past the initial slab capacity and hash
  // bucket count. If growth fails, we would crash or get nullptr derefs.
  std::vector<Bdd> arena;
  arena.reserve(200);
  for (int i = 0; i < 200; ++i) {
    auto a = getVar(i % varNum);
    auto b = getVar((i + 1) % varNum);
    auto c = getVar((i + 2) % varNum);
    arena.push_back((a ^ b) | (a & c));
  }

  assert(numNodes() > baseline + 10 &&
         "Slab growth / maintenance path should allow many allocations");

  // Release BDDs so their roots are deref'ed, then run GC to reclaim.
  arena.clear();
  gc();

  const auto afterGc = numNodes();
  const auto maxReachable = 2 + 2 * varNum; // terminals + vars + nvars
  assert(afterGc <= maxReachable &&
         "GC should reclaim unreachable nodes into free list");

  // Allocate again; should succeed via free list without unbounded growth.
  const auto beforeAlloc = numNodes();
  auto d = getVar(0) ^ getVar(1);
  (void)d;
  assert(numNodes() <= beforeAlloc + 4 &&
         "Re-allocation should use reclaimed nodes or a small growth");

  std::cout << "node_table_slab_test passed" << std::endl;
  return 0;
}

File: rust/tests/node_table_slab_test.rs
use std::sync::Mutex;

use nanobdd::{gc, init, num_nodes, var};

static TEST_MUTEX: Mutex<()> = Mutex::new(());

#[test]
fn node_table_slab_growth_and_gc() {
  let _guard = TEST_MUTEX.lock().expect("test mutex poisoned");

  // Ensure a clean global state before starting.
  unsafe { nanobdd::clear() };

  const TABLE_SIZE: usize = 5;
  const CACHE_SIZE: usize = 5;
  const VAR_NUM: usize = 3;

  init(TABLE_SIZE, CACHE_SIZE, VAR_NUM);

  let baseline = num_nodes();

  // Trigger many unique nodes to push past the initial slab capacity and hash
  // bucket count. If growth fails, we'd crash or get null derefs.
  let mut arena = Vec::with_capacity(200);
  for i in 0..200 {
    let a = var((i % VAR_NUM) as u32);
    let b = var(((i + 1) % VAR_NUM) as u32);
    let c = var(((i + 2) % VAR_NUM) as u32);
    arena.push(&(&a ^ &b) | &(&a & &c));
  }

  assert!(
    num_nodes() > baseline + 10,
    "slab growth / maintenance path should allow many allocations"
  );

  // Release BDDs so their roots are deref'ed, then run GC to reclaim.
  arena.clear();
  gc();

  let after_gc = num_nodes();
  let max_reachable = 2 + 2 * VAR_NUM; // terminals + vars + nvars
  assert!(
    after_gc == max_reachable,
    "GC should reclaim unreachable nodes into free list"
  );

  // Allocate again; should succeed via free list without unbounded growth.
  let before_alloc = num_nodes();
  let d = {
    let v0 = var(0);
    let v1 = var(1);
    &v0 ^ &v1
  };
  drop(d);
  assert!(
    num_nodes() == before_alloc + 1,
    "re-allocation should use reclaimed nodes or a small growth"
  );

  unsafe { nanobdd::clear() };
}
